src/transformers/trainer_tf.py

```diff
@@ -0,0 +1,509 @@
+        """
+        Setup the different steps to train a model:
+          - check if all the data are given
+          - create the proper strategy
+          - create the features
+          - prepare the model settings
+
+        Args:
+          checkpoint_path: the directory path where the model checkpoints will be saved.
+          log_path: the directory path where the Tensorboard logs will be saved.
+          cache_dir (optional): the directory path where the pretrained model and data will be cached.
+            "./cache" folder by default.
+          data_dir (optional): the directoty path where the data are. This parameter becomes mandatory if
+            the parameters training_data and validation_data are not given.
+          training_data (optional): the training data, either as a list of example or as a tf.data.Dataset.
+            This parameter becomes mandatory if the parameter data_dir is not given.
+          validation_data (optional): the validation data, either as a list of example or as a tf.data.Dataset.
+            This parameter becomes mandatory if the parameter data_dir is not given.
+          test_data (optional): the test data, either as a list of example or as a tf.data.Dataset.
+            This parameter will take the validation data as value if nothing is given.
+        """
+        if self.distributed:
+            self.strategy = tf.distribute.MirroredStrategy()
+            self.batch_size *= self.strategy.num_replicas_in_sync
+            self.eval_batch_size *= self.strategy.num_replicas_in_sync
+        else:
+            if len(tf.config.list_physical_devices('GPU')) >= 1:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/gpu:0")
+            else:
+                self.strategy = tf.distribute.OneDeviceStrategy(device="/cpu:0")
+
+        tokenizer = AutoTokenizer.from_pretrained(self.pretrained_model_name_or_path, cache_dir=cache_dir, use_fast=False)
+
+        self._preprocessing_data(tokenizer, cache_dir)
+
+        if len(self.labels) == 0:
+            self.labels = self.processor.get_labels()
+
+        self.label2id = {label: i for i, label in enumerate(self.labels)}
+        self.id2label = {i: label for i, label in enumerate(self.labels)}
+        config = AutoConfig.from_pretrained(self.pretrained_model_name_or_path, num_labels=len(self.labels), id2label=self.id2label, label2id=self.label2id, cache_dir=cache_dir)
+
+        with self.strategy.scope():
+            self.model = self.architecture.from_pretrained(self.pretrained_model_name_or_path, config=config, cache_dir=cache_dir)
+            self._create_optimizer()
+            self._get_loss_and_metric()
+            self._create_checkpoint_manager(checkpoint_path)
+            self._create_summary_writer(log_path)
+
+    def get_labels(self):
+        """
+        Returns the list of labels associated to the trained model.
+        """
+        return self.labels
+
+    def _get_loss_and_metric(self):
```

Comment:
Should we call it `_set_loss_and_metric` since it doesn't return anything?

Reply:
Good catch!